File: FlutterCode/lib/domain/entities/invitation.dart
import 'package:uuid/v8.dart';

enum InvitationStatus {
  pending,
  accepted,
  declined,
}

class Invitation {
  final String id;
  final String inviteeUserId;
  final String inviterUserId;
  final String familyId;
  final bool isParent;
  final DateTime sentAt;
  final InvitationStatus status;

  Invitation._({
    required this.id,
    required this.inviteeUserId,
    required this.inviterUserId,
    required this.familyId,
    required this.isParent,
    required this.sentAt,
    required this.status,
  });

  static final UuidV8 _uuidGenerator = UuidV8();

  static Invitation create({
    required String inviteeUserId,
    required String inviterUserId,
    required String familyId,
    required bool isParent,
  }) {
    return Invitation._(
      id: _uuidGenerator.generate(),
      inviteeUserId: inviteeUserId,
      inviterUserId: inviterUserId,
      familyId: familyId,
      isParent: isParent,
      sentAt: DateTime.now(),
      status: InvitationStatus.pending,
    );
  }

  Invitation accept() {
    if (status == InvitationStatus.accepted) {
      throw StateError("Invitation already accepted.");
    }
    if (status == InvitationStatus.declined) {
      throw StateError("Invitation was declined and cannot be accepted.");
    }

    return Invitation._(
      id: id,
      inviteeUserId: inviteeUserId,
      inviterUserId: inviterUserId,
      familyId: familyId,
      isParent: isParent,
      sentAt: sentAt,
      status: InvitationStatus.accepted,
    );
  }

  Invitation decline() {
    if (status == InvitationStatus.declined) {
      throw StateError("Invitation already declined.");
    }
    if (status == InvitationStatus.accepted) {
      throw StateError("Invitation was accepted and cannot be declined.");
    }

    return Invitation._(
      id: id,
      inviteeUserId: inviteeUserId,
      inviterUserId: inviterUserId,
      familyId: familyId,
      isParent: isParent,
      sentAt: sentAt,
      status: InvitationStatus.declined,
    );
  }

  Invitation cancel({required String requesterUserId}) {
    if (requesterUserId != inviterUserId) {
      throw StateError("Only the inviter can cancel this invitation.");
    }
    if (status != InvitationStatus.pending) {
      throw StateError("Only pending invitations can be canceled.");
    }

    return Invitation._(
      id: id,
      inviteeUserId: inviteeUserId,
      inviterUserId: inviterUserId,
      familyId: familyId,
      isParent: isParent,
      sentAt: sentAt,
      status: InvitationStatus.declined,
    );
  }

  static Invitation fromId({
    required String id,
    required String inviteeUserId,
    required String inviterUserId,
    required String familyId,
    required bool isParent,
    required DateTime sentAt,
    required InvitationStatus status,
  }) {
    return Invitation._(
      id: id,
      inviteeUserId: inviteeUserId,
      inviterUserId: inviterUserId,
      familyId: familyId,
      isParent: isParent,
      sentAt: sentAt,
      status: status,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      "id": id,
      "inviteeUserId": inviteeUserId,
      "inviterUserId": inviterUserId,
      "familyId": familyId,
      "isParent": isParent,
      "sentAt": sentAt.toIso8601String(),
      "status": status.name,
    };
  }

  static Invitation fromJson(Map<String, dynamic> json) {
    return Invitation._(
      id: json["id"],
      inviteeUserId: json["inviteeUserId"],
      inviterUserId: json["inviterUserId"],
      familyId: json["familyId"],
      isParent: json["isParent"],
      sentAt: DateTime.parse(json["sentAt"]),
      status: InvitationStatus.values.byName(json["status"]),
    );
  }
}

File: backend/Expense_Tracker.Domain/Invitation/Invitation.cs
using ErrorOr;
using Expense_Tracker.Domain.Common;
using Expense_Tracker.Domain.Errors;
using Expense_Tracker.Domain.FamilyFolder;
using Expense_Tracker.Domain.Invitation.Enums;

namespace Expense_Tracker.Domain.Invitation;

public sealed class Invitation : Entity
{
    public Guid InviteeUserId { get; private set; }
    public Guid InviterUserId { get; private set; }
    public Guid FamilyId { get; private set; }
    public bool IsParent { get; private set; }
    public DateTimeOffset SentAtUtc { get; private set; }
    public InvitationStatus Status { get; private set; }

    public Family Family { get; private set; } = null!;

    private Invitation() { }

    private Invitation(
        Guid id,
        Guid inviteeUserId,
        Guid inviterUserId,
        Guid familyId,
        bool isParent) : base(id)
    {
        InviteeUserId = inviteeUserId;
        InviterUserId = inviterUserId;
        FamilyId = familyId;
        IsParent = isParent;
        SentAtUtc = DateTimeOffset.UtcNow;
        Status = InvitationStatus.Pending;
    }

    public static ErrorOr<Invitation> Create(
        Guid inviteeUserId,
        Guid inviterUserId,
        Guid familyId,
        bool isParent)
    {
        if (inviteeUserId == Guid.Empty)
            return DomainErrors.GeneralErrors.InvalidState(nameof(Invitation), "Invitee user ID is required.");

        if (inviterUserId == Guid.Empty)
            return DomainErrors.GeneralErrors.InvalidState(nameof(Invitation), "Inviter user ID is required.");

        if (familyId == Guid.Empty)
            return DomainErrors.GeneralErrors.InvalidState(nameof(Invitation), "Family ID is required.");

        if (inviteeUserId == inviterUserId)
            return DomainErrors.InvitationErrors.SelfInvite();

        var invitation = new Invitation(
            Guid.CreateVersion7(),
            inviteeUserId,
            inviterUserId,
            familyId,
            isParent);

        return invitation;
    }

    public ErrorOr<Success> Accept()
    {
        if (Status == InvitationStatus.Accepted)
            return DomainErrors.InvitationErrors.AlreadyAccepted();

        if (Status == InvitationStatus.Declined)
            return DomainErrors.InvitationErrors.AlreadyDeclined();

        if (Status == InvitationStatus.Cancelled)
            return DomainErrors.InvitationErrors.Cancelled();

        Status = InvitationStatus.Accepted;
        return new Success();
    }

    public ErrorOr<Success> Decline()
    {
        if (Status == InvitationStatus.Declined)
            return DomainErrors.InvitationErrors.AlreadyDeclined();

        if (Status == InvitationStatus.Accepted)
            return DomainErrors.GeneralErrors.InvalidState(nameof(Invitation), "Invitation was accepted and cannot be declined.");

        if (Status == InvitationStatus.Cancelled)
            return DomainErrors.InvitationErrors.Cancelled();

        Status = InvitationStatus.Declined;
        return new Success();
    }

    public ErrorOr<Success> Cancel(Guid requesterId)
    {
        if (Status != InvitationStatus.Pending)
            return DomainErrors.InvitationErrors.NotPending();

        Status = InvitationStatus.Cancelled;
        return new Success();
    }
}
